src/http/Server.cpp

```diff
@@ -0,0 +1,343 @@
+    QString tmp_out = QString(*out);
+    int tmp_pos1 = tmp_out.indexOf("\"Count\":0,");
+    int tmp_pos2 = tmp_out.indexOf("\"Entries\":null,");
+    if (tmp_pos1 != -1 && tmp_pos2 != -1) {
+        tmp_out.replace(tmp_pos2, 15, "\"Entries\":[],");
+    }
+    *out = tmp_out.toUtf8();
+
+    Q_EMIT donewrk();
+}
+
+
+void Server::handleOpenDatabase(bool *success)
+{
+    *success = openDatabase();
+    Q_EMIT donewrk();
+}
+
+
+int Server::request_handler(struct MHD_Connection *connection,
+        const char *, const char *method, const char *,
+        const char *upload_data, size_t *upload_data_size, void **con_cls)
+{
+    struct Server::connection_info_struct *con_info =
+        static_cast<struct Server::connection_info_struct*>(*con_cls);
+
+    if (!isDatabaseOpened()) {
+        bool success;
+        QEventLoop loop1;
+        loop1.connect(this, SIGNAL(donewrk()), SLOT(quit()));
+        Q_EMIT emitOpenDatabase(&success);
+        loop1.exec();
+
+        if (!success)
+            return send_unavailable(connection);
+    }
+
+    if (con_info == NULL) {
+        *con_cls = calloc(1, sizeof(*con_info));
+        return MHD_YES;
+    }
+
+    if (strcmp (method, MHD_HTTP_METHOD_POST) != 0)
+        return MHD_NO;
+
+    if (*upload_data_size == 0) {
+        if (con_info && con_info->response)
+            return send_response(connection, con_info->response);
+        else
+            return MHD_NO;
+    }
+
+    QString type = MHD_lookup_connection_value(connection,
+                                               MHD_HEADER_KIND, "Content-Type");
+    if (!type.contains("application/json", Qt::CaseInsensitive))
+        return MHD_NO;
+
+    // Now process the POST request
+
+    QByteArray post = QByteArray(upload_data, *upload_data_size);
+
+    QByteArray s;
+    QEventLoop loop;
+    loop.connect(this, SIGNAL(donewrk()), SLOT(quit()));
+    Q_EMIT emitRequest(post, &s);
+    loop.exec();
+
+    if (s.size() == 0)
+        return MHD_NO;
+
+    con_info->response = static_cast<char*>(calloc(1, s.size()+1));
+    memcpy(con_info->response, s.data(), s.size());
+
+    *upload_data_size = 0;
+
+    return MHD_YES;
+}
+
+
+void Server::request_completed(void *, struct MHD_Connection *,
+        void **con_cls, enum MHD_RequestTerminationCode)
+{
+    struct Server::connection_info_struct *con_info =
+        static_cast<struct Server::connection_info_struct*>(*con_cls);
+
+    if (con_info == NULL)
+        return;
+
+    if (con_info->response) free(con_info->response);
+    free(con_info);
+    *con_cls = NULL;
+}
+
+
+void Server::start(void)
+{
+    if (m_started) return;
+
+    int port = HttpSettings::httpPort();
+
+    daemon = MHD_start_daemon(MHD_USE_SELECT_INTERNALLY, port, NULL, NULL,
+                              &this->request_handler_wrapper, this,
+                              MHD_OPTION_NOTIFY_COMPLETED,
+                              this->request_completed, NULL,
```

Comment:
This will bind 0.0.0.0 and will listen at all requests made to KeePassX.

This is a security flaw and an adversary can spoof your key and request almost all your passwords.
PLEASE, DON'T DO THIS.